allennlp/nn/util.py

```diff
@@ -402,9 +403,99 @@ def inner_device_mapping(storage: torch.Storage, location) -> torch.Storage:  #
             return storage
     return inner_device_mapping
 
+
 def ones_like(tensor: torch.Tensor) -> torch.Tensor:
     """
     Use clone() + fill_() to make sure that a ones tensor ends up on the right
     device at runtime.
     """
     return tensor.clone().fill_(1)
+
+
+def combine_tensors(combination: str, tensors: List[torch.Tensor]) -> torch.Tensor:
+    """
+    Combines a list of tensors using element-wise operations and concatenation, specified by a
+    ``combination`` string.  The string refers to (1-indexed) positions in the input tensor list,
+    and looks like ``"1,2,1+2,3-1"``.
+
+    We allow the following kinds of combinations: ``x``, ``x*y``, ``x+y``, ``x-y``, and ``x/y``,
+    where ``x`` and ``y`` are positive integers less than or equal to ``len(tensors)``.  Each of
+    the binary operations is performed elementwise.  You can give as many combinations as you want
+    in the ``combination`` string.  For example, for the input string ``"1,2,1*2"``, the result
+    would be ``[1;2;1*2]``, as you would expect, where ``[;]`` is concatenation along the last
+    dimension.
+
+    If you have a fixed, known way to combine tensors that you use in a model, you should probably
+    just use something like ``torch.cat([x_tensor, y_tensor, x_tensor * y_tensor])``.  This
+    function adds some complexity that is only necessary if you want the specific combination used
+    to be `configurable`.
+
+    If you want to do any element-wise operations, the tensors involved in each element-wise
+    operation must have the same shape.
+
+    This function also accepts ``x`` and ``y`` in place of ``1`` and ``2`` in the combination
+    string.
+    """
+    combination = combination.replace('x', '1').replace('y', '2')
+    combinations = combination.split(',')
+    combined_tensor = _get_combination(combinations[0], tensors)
+    for piece in combinations[1:]:
+        to_concatenate = _get_combination(piece, tensors)
+        combined_tensor = torch.cat([combined_tensor, to_concatenate], dim=-1)
```

Comment:
is it inefficient to concatenate tensors one at a time rather than all at once?  (I don't know)

Reply:
Yeah, that's a good point.  I don't know why I did it this way originally...  I'll fix it.